exploration/data-driven-tests.md

```diff
@@ -0,0 +1,243 @@
+  }
+  // "verify":  "{ formatter: "string", value: "foo" } and { formatter: number, value: 123 }"
+}
+```
+
+#### Formatter tests
+
+These tests focus on the standard registry's formatters (e.g. `:number`, `:datetime`). They cover the different options that can be passed to each formatter (e.g. `offset`, `skeleton`).
+
+Example:
+
+```jsonc
+{
+  "label": "Skeleton affects datetime format",
+  "locale": "en-US",
+  "pattern": "{$givenDateTime :datetime skeleton=yMMMdE}",
+  "inputs": {
+    "givenDateTime": { "type": "datetime", "value": "2000-12-31T00:00:00.000Z" }
+  }
+  // "verify":  "Sun, 31 Dec 2000"
+}
+```
+
+#### Selector tests
+
+These are extensive tests of the cases within a `match` statement. Testing of multiple selectors is included.
+
+Example:
+
+```jsonc
+{
+  "label": "Matches numbers other than one",
+  "locale": "en-US",
+  "pattern": "match {$arg :number} when one {one} when * {*}",
+  "inputs": {
+    "arg": { "type": "number", "value": 3 }
+  }
+  // "verify": "*"
+}
+```
+
+#### Data model tests (optional)
+
+There is no standard data model within the specification, which means that we cannot create mandatory data model tests.
+
+If a particular implementation of MF2 exposes a standardized representation of [the data model](../spec/data-model/message.json), perhaps through a `mf2.toCanonicalJson();` function or similar, then we could create tests that assert against this.
+
+## Alternatives Considered
+
+### Gherkin syntax and Cucumber runner
+
+Based on the readability concerns mentioned above, the Gherkin syntax was also considered.
+
+Example:
+
+```feature
+Feature: Multi-selector messages
+
+  Background:
+    Given the username is "Matt"
+    And the source is:
+      """
+      match {$photoCount :number} {$userGender :equals}
+      when 1 masculine {{$userName} added a new photo to his album.}
+      when 1 feminine  {{$userName} added a new photo to her album.}
+      when 1 *         {{$userName} added a new photo to their album.}
+      when * masculine {{$userName} added {$photoCount} photos to his album.}
+      when * feminine  {{$userName} added {$photoCount} photos to her album.}
+      when * *         {{$userName} added {$photoCount} photos to their album.}
+      """
+
+  Scenario: One item - male
+    When the message is resolved with params:
+      | key        | value     |
+      | photoCount |         1 |
+      | userGender | masculine |
+    Then the string output is "Matt added a new photo to his album."
+```
+
+The [Cucumber framework](https://cucumber.io/) was considered because of its integration with the Gherkin syntax. Cucumber's approach of using platform-specific step definitions for Gherkin scenarios aligns with our goal of having a data-only representation of the test content. Unfortunately, the MF2 test runner may need to target a wide range of platforms in future. In this case, integrating with Cucumber would incur additional overhead.
```

Comment:
At the risk of repeating @stasm from above, another challenge could be integrating cucumber with existing, possibly even legacy platforms.

Reply:
Agreed. I'll reword accordingly.